lib/Basics/VelocyPackHelper.cpp

```diff
@@ -387,6 +391,188 @@ int VelocyPackHelper::compareNumberValues(VPackValueType lhsType,
+  if (a == b) {
+    return 0;
+  }
+  return (a < b ? -1 : 1);
+}
+
+static int compareInt64UInt64(int64_t i, uint64_t u) {
+  if (i < 0) {
+    return -1;
+  }
+  return comp<uint64_t>(static_cast<uint64_t>(i), u);
+}
+
+constexpr uint64_t uint64_2_52 = uint64_t{1} << 52;
+constexpr uint64_t uint64_2_63 = uint64_t{1} << 63;
+constexpr double double_2_52 = static_cast<double>(uint64_2_52);
+const double double_2_64 = ldexp(1.0, 64);
+
+static int compareUInt64Double(uint64_t u, double d) {
+  if (std::isnan(d)) {
+    return -1;
+  }
+  if (d < 0.0) {
+    return 1;
+  }
+  if (u < uint64_2_52) {
+    return comp<double>(static_cast<double>(u), d);
+  }
+  if (d < double_2_52) {
+    return 1;  // since we know that u >= 2^52
+  }
+  if (d >= double_2_64) {
+    // Note that this covers the case that d = +inf !
+    return -1;  // since we know that u < 2^64
+  }
+  // Now we know that both numbers are >= 2^52 and < 2^64, so we can look
+  // at the integral and fractional parts of d and decide then:
+  double df = floor(d);
+  uint64_t du = static_cast<uint64_t>(df);  // will work and give the
+                                            // correct integer!
+  if (u < du) {
+    return -1;
+  }
+  if (u > du) {
+    return 1;
+  }
+  // So u == du, let's check the fractional part of d:
+  if (d == df) {
+    // No fractional part, so d and u are equal:
+    return 0;
+  }
+  // There is a fractional part, so u is smaller:
+  return -1;
+}
+
+static int compareInt64Double(int64_t i, double d) {
+  if (std::isnan(d)) {
+    return -1;
+  }
+  if (i < 0) {
+    uint64_t u =
+        (i == std::numeric_limits<int64_t>::min() ? uint64_2_63
+                                                  : static_cast<uint64_t>(-i));
+    return -compareUInt64Double(u, -d);
+  }
+  return compareUInt64Double(static_cast<uint64_t>(i), d);
+}
+
+int VelocyPackHelper::compareNumberValues(VPackValueType lhsType,
+
+                                          VPackSlice lhs, VPackSlice rhs) {
+  if (lhsType == rhs.type()) {
+    // both types are equal
+    if (lhsType == VPackValueType::Int || lhsType == VPackValueType::SmallInt) {
+      // use exact comparisons. no need to cast to double
+      int64_t l = lhs.getIntUnchecked();
+      int64_t r = rhs.getIntUnchecked();
+      return comp<int64_t>(l, r);
+    }
+
+    if (lhsType == VPackValueType::UInt) {
+      // use exact comparisons. no need to cast to double
+      uint64_t l = lhs.getUIntUnchecked();
+      uint64_t r = rhs.getUIntUnchecked();
+      return comp<uint64_t>(l, r);
+    }
+
+    if (lhsType == VPackValueType::Double) {
+      double l = lhs.getDouble();
+      double r = rhs.getDouble();
+      if (std::isnan(l)) {
+        if (std::isnan(r)) {
+          return 0;
+        }
+        return 1;
+      } else if (std::isnan(r)) {
+        return -1;
+      }
+      // No nan on either side!
+      return comp<double>(l, r);
+    }
+  }
+
+  VPackValueType rhsType = rhs.type();
```

Comment:
You could put this above the previous `if (lhsType == rhs.type())`, and avoid the second `.type()` call.

Reply:
done.